### DISCUSSION TITLE
Inconsistent message rate

### DISCUSSION THREAD

**Discussion by arcarmona** [emqx/emqx#12502] [Q&A] [answered]
Hello,

I have currently set up an emqx (community edition) cluster in kubernetes using emqx-operator , I have around 300 clients and an expected message income of ~6k/s (message size ~1Kb). When running I get the expected connections, however when the message rate seems to be setting to its expected number, it drops (emqx dashboards also lags a bit before the drop).

I understand there can be small oscillations in the message income however for my use case such a big drop shouldnt be happening and the message rate per second should stay consistent around 6k. I wanted to ask if I might be missing some configuration parameter that might help solve this problem. Attached below you may find an screenshot of the behavior and my emqx yaml.

Thank you.

![image](https://github.com/emqx/emqx/assets/152393447/71bd54b3-acb6-43e7-b756-3e96085050fd)

```
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
spec:
  image: emqx:5.4.1
  imagePullPolicy: Always
  updateStrategy:
    evacuationStrategy:
      connEvictRate: 1000
      sessEvictRate: 1000
      waitTakeover: 10
    initialDelaySeconds: 10
    type: Recreate
  config:
    data: |
      listeners.tcp.default {
        proxy_protocol = true,
        max_connections = 10000,
        acceptors = 64,
        messages_rate = "10000/s",
        max_conn_rate = "10000/s"
      }
      listeners.ssl.default {
        enable = false
      }
      mqtt {
        max_topic_levels = 20,
        peer_cert_as_username = "cn",
        shared_subscription_strategy = "random",
        ignore_loop_deliver = true,
        max_inflight = 1000
      }
      node {
        dist_buffer_size = 2097150,
      }
      force_shutdown {
        max_heap_size = "64MB"
      }
      retainer {
        msg_expiry_interval = "10s",
        msg_clear_interval = "10s"
      }
  coreTemplate:
    spec:
      resources:
        limits:
          cpu: "6"
          memory: 6Gi
        requests:
          cpu: "6"
          memory: 6Gi
      ports:
        - name: dashboard
          containerPort: 18083
      replicas: 2
      volumeClaimTemplates:
        storageClassName: standard
        resources:
          requests:
            storage: 10Gi
        accessModes:
        - ReadWriteOnce
  dashboardServiceTemplate:
    spec:
      type: LoadBalancer
  listenersServiceTemplate:
    spec:
      clusterIP: None 
```

**Comment by keynslug** [comment]
Hi @arcarmona,

It's very hard to tell what's going on without a lot of additional information. I'd say having these bits of information is a good start:
1. What are the clients? How do you make sure they are pushing 6k msg/s and not exceeding this rate?
2. Is there enough computing resources? What are CPU / memory utilization on the instance running EMQX?
3. Is there anything in the EMQX logs worth sharing (i.e. warnings or errors)?

If you could share workload specification in some reproducible form (i.e. a script), it would be perfect.

**Reply by arcarmona** [reply]
Hello @keynslug, first of all thank you for your swift response. To add more detail to the use case:

1. The clients are bridged devices which from previous experience with mosquitto we know amount to ~6k msgs/s. It seems key to know as well that these devices together carry approximately 2M retained messages (I believe this was the source of the issue)
2. We are deploying in GKE now with 3 core nodes (6 cpus and 6G RAM each). Resources seem to suffice as usage does not seem to come close to these limits.
3. When checking logs what stood out the most were the following warnings:

     - Mnesia is overloaded: {dump_log,time_threshold}, “,”level”:”warning”,”error_logger”:{“tag”:”warning_msg”}}
     - [error] supervisor: ‘esockd_connection_sup - <0.2772.0>’, errorContext: connection_shutdown, reason: #{max => 1000,reason => message_queue_too_long,value => 1410}. 

Upon searching online, I found the latter warning could be fixed increasing force_shutdown.max_mailbox_size, I went ahead and just diabled force_shutdown altogether with force_shutdown.enable = false. I redeployed and after some time without any changes I disabled retained messages and this fixed it(income is as it should and works perfectly).

It is most likely we will move forward with this setup (no retain), however I would like to know if possible what was happening before which I assume had to do with the 2M retained messages. In addition, in case you find useful, once I disabled retained (after having it on a while), the income rate increased drastically for a short amount of time reaching even 70k/s msgs, but shortly after it stabilized at the expected throughput.

As for how the workload was set up, I followed emqx-operator docs on deployment on GKE and setup a haproxy load balancer in front of it.

**Reply by keynslug** [reply, marked as answer]
Thanks for the details.

> 1. It seems key to know as well that these devices together carry approximately 2M retained messages (I believe this was the source of the issue)

Sounds highly likely actually. This is quite a lot for the current retainer implementation. If these retained messages' topics and respective subscriptions follow any predictable patterns, it could be possible to tune the retainer configuration, however the possibilities are somewhat limited.

> In addition, in case you find useful, once I disabled retained (after having it on a while), the income rate increased drastically for a short amount of time reaching even 70k/s msgs, but shortly after it stabilized at the expected throughput.

The following explanation sounds plausible: the moment you turned off the retainer it stopped being the bottleneck in the message processing, and all the messages that were queued in respective connection processes quickly passed through (there could have been thousands of them per each connection, according to the warnings in the logs you've seen before).

**Reply by arcarmona** [reply]
Thanks for the answer! It confirms my initial thoughts, I will mark this as answered
